### PR TITLE
Restructure LIBDIR: Move Dynlink, Str and Unix to sub-directories (port PR#11198)

### PR DESCRIPTION
**Pull request by gretay-js**
Port of part of https://github.com/ocaml/ocaml/pull/11198 that deals with Makefiles, and similar changes to dune files. 

~Still a draft,  `make runtest-upstream` fails when `ocamltest` refers to these libraries, doesn't find them.~

Finally, github CI is green on all but 32-bit build. 

Ready for review.

This PR now includes backporting https://github.com/ocaml-flambda/flambda-backend/pull/895 which adds `Clflags.no_cwd` that is used in one place in ocaml#11198.

### CODE REVIEW

**Review by stedolan — APPROVED** [review]
One typo, but otherwise looks good to me!

File: ocaml/otherlibs/str/dune [hunk 1]
```diff
@@ -29,17 +29,17 @@
...
-    str.mli
-    .str.objs/byte/str.cmi
-    .str.objs/byte/str.cmti
+    (str.cmxa as str/str.cmxa)
+    (str.a as str/str.a)
+    (str.cmxs as str/str.cmxs)
+    (str.cma as str/str.cma)
+    (str.mli as str/str.mli)
+    (.str.objs/byte/str.cmi as str/str.cmi)
+    (.str.objs/byte/str.cmti as tr/str.cmti)
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply ocaml/otherlibs/str/dune]: fixed. thank you!

**Comment by lthls** [comment]
This PR breaks the opam builds because dune and ocamlfind still expect the original directory structure. I would advise reverting it until we've found a good way to make the tools aware that we're using the same directory structure as OCaml 5.0.

**Comment by lpw25** [comment]
I'd rather avoid reverting it. We're switching our build system to look in the new place, and it will not be easy to switch it back. Can't you just add a step to symlink them from the old places as part of the opam install instructions?

**Comment by lthls** [comment]
That's one possibility. One other might be to ask David for all the external tools that he had to patch, and find a way to use their 5.0-compatible versions instead of their 4.14 versions.

**Comment by gretay-js** [comment]
maybe another way is to make a follow PR to turn the alert into a "proper" warning (improving location would be nice) and then disable the warning by default when building on opam.

**Comment by lthls** [comment]
I don't think the warnings are that problematic (although of course we might want to fix them in some way). The build failures are real errors, that wouldn't disappear if we removed the warnings. If I understand correctly, dune detects that the unix library (and the other ones, of course) is missing and will fail to compile any library that depends on it. And there are similar errors with ocamlfind. Projects with hand-crafted build systems will see the alerts, but compile fine otherwise.

**Comment by gretay-js** [comment]
I thought that the missing directory would be added by default by the compiler whenever the warning/alert is issued. 
oh, maybe it's the case of transitive dependencies that needs fixing?

**Comment by lthls** [comment]
If the build system actually tried to compile the files, it would work. But the build system assumes that the libraries are missing and does not even try.

**Pull request merged**